Max cap on partition movement rate during executions

smtwilio opened the pull request:
This PR resolves https://github.com/linkedin/cruise-control/issues/1725

### Context:
At Twilio, we run hundreds of Kafka brokers spread across multiple clusters, environments, and regions and are an integral component to many of Twilio’s core products. 

This is an observation from the last production deployment. Right now, we artificially limit the disruption to clients on the fleet during deployment by doing, say, 20 brokers at a time, with the CC setting of # of concurrent partition movements per broker.

It might be better for us to extend CC so that we can set an overall cap of partition movements per reassignment so that we wouldn't need to do the 20 brokers at a time thing. A single CC operation (all brokers in the fleet) would be fine since the underlying step-by-step CC process would make sure that not too many OVERALL partitions were being moved at once.

### Changes
Changes are in 3 parts - 3 commits
1. Core change: Added the config `max.num.cluster.partition.movements` with default 1200 and logic to throttle the tasks being returned to the executor - 
`cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java`
2. Rest argument and dynamic update of parameter: This is 2 parts again. 
  a) Rest Parameters updated for the sync actions AddOrRemoveBrokers, Rebalance, FixOfflineReplicas, and TopicReplicationFactor change along with async action ChangeConcurrencyParameters
   b) Admin operation to update the value during runtime
3. Test cases updated 

### Usage Example
#### Removing a broker
Find `max_partition_movements_in_cluster=8` in command below
```
/kafkacruisecontrol/remove_broker?brokerid=9006688&destination_broker_ids=8095530&concurrent_partition_movements_per_broker=30&max_partition_movements_in_cluster=8&concurrent_leader_movements=30&json=true&verbose=true&json=true&dryrun=true' 
```

#### Updating the value during runtime
Updating the value to 5 
```
$ curl -X POST '<SERVER>/kafkacruisecontrol/admin?max_partition_movements_in_cluster=5'
{
ongoingConcurrencyChangeRequest: Max Inter-broker partition movements is set to 5
}
```

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/ChangeExecutionConcurrencyParameters.java:
```diff
@@ -29,13 +30,15 @@
     validParameterNames.add(CONCURRENT_PARTITION_MOVEMENTS_PER_BROKER_PARAM);
     validParameterNames.add(CONCURRENT_INTRA_BROKER_PARTITION_MOVEMENTS_PARAM);
     validParameterNames.add(CONCURRENT_LEADER_MOVEMENTS_PARAM);
+    validParameterNames.add(MAX_PARTITION_MOVEMENTS_IN_CLUSTER_PARAM);
```
Can we update the JavaDoc of `AdminParameters` to show the use of this new parameter?

smtwilio replied:
Done!

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/ChangeExecutionConcurrencyParameters.java:
```diff
@@ -87,6 +92,9 @@ public Integer concurrentLeaderMovements() {
 
+  public Integer maxInterBrokerPartitionMovements() {
+    return _maxInterBrokerPartitionMovements;
+  }
```
Nit: missing new line after function

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/FixOfflineReplicasParameters.java:
```diff
@@ -53,6 +54,7 @@
     validParameterNames.add(DRY_RUN_PARAM);
     validParameterNames.add(REASON_PARAM);
     validParameterNames.add(CONCURRENT_PARTITION_MOVEMENTS_PER_BROKER_PARAM);
+    validParameterNames.add(MAX_PARTITION_MOVEMENTS_IN_CLUSTER_PARAM);
```
Can we update the JavaDoc of this class to show the use of this new parameter?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/RebalanceParameters.java:
```diff
@@ -57,6 +58,7 @@
     validParameterNames.add(DRY_RUN_PARAM);
     validParameterNames.add(REASON_PARAM);
     validParameterNames.add(CONCURRENT_PARTITION_MOVEMENTS_PER_BROKER_PARAM);
+    validParameterNames.add(MAX_PARTITION_MOVEMENTS_IN_CLUSTER_PARAM);
```
Can we update the JavaDoc of this class to show the use of this new parameter?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/TopicReplicationFactorChangeParameters.java:
```diff
@@ -36,6 +37,7 @@
     validParameterNames.add(REPLICATION_FACTOR_PARAM);
     validParameterNames.add(SKIP_RACK_AWARENESS_CHECK_PARAM);
     validParameterNames.add(CONCURRENT_PARTITION_MOVEMENTS_PER_BROKER_PARAM);
+    validParameterNames.add(MAX_PARTITION_MOVEMENTS_IN_CLUSTER_PARAM);
```
Can we update the JavaDoc of `TopicConfigurationParameters` to show the use of this new parameter?

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/config/ConcurrencyConfigTest.java:
```diff
@@ -266,6 +284,8 @@ public void testMinExecutionProgressCheckIntervalMsGreaterThanExecutionProgressC
     EasyMock.expect(config.getInt(ExecutorConfig.MAX_NUM_CLUSTER_MOVEMENTS_CONFIG))
             .andReturn(ExecutorConfig.DEFAULT_MAX_NUM_CLUSTER_MOVEMENTS_CONFIG);
+    EasyMock.expect(config.getInt(ExecutorConfig.MAX_NUM_CLUSTER_PARTITION_MOVEMENTS_CONFIG))
+            .andReturn(ExecutorConfig.DEFAULT_MAX_NUM_CLUSTER_PARTITION_MOVEMENTS_CONFIG);
```
Should we add a new test case to this class to test the case for the new constraint added by this PR?

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManagerTest.java:
```diff
@@ -522,7 +525,7 @@ public void testExecutionInProgress() throws InterruptedException {
     Properties props = KafkaCruiseControlUnitTestUtils.getKafkaCruiseControlProperties();
     KafkaCruiseControlConfig kafkaCruiseControlConfig = new KafkaCruiseControlConfig(props);
-    EasyMock.expect(mockKafkaCruiseControl.config()).andReturn(kafkaCruiseControlConfig).times(2);
+    EasyMock.expect(mockKafkaCruiseControl.config()).andReturn(kafkaCruiseControlConfig).atLeastOnce();
```
Can we make the number of calls specific?

smtwilio replied:
Reverted this change in all files

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/IdempotenceCacheTest.java:
```diff
@@ -74,7 +74,7 @@ public void testIdempotence() {
       parameterConfigOverrides.put(MAINTENANCE_EVENT_TYPE_CONFIG, eventType);
       // Expect mocks.
-      EasyMock.expect(_mockKafkaCruiseControl.config()).andReturn(_config).times(4);
+      EasyMock.expect(_mockKafkaCruiseControl.config()).andReturn(_config).atLeastOnce();
```
(The same comment applies to other similar uses) can we avoid dropping `times(...)`?

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/IdempotenceCacheTest.java:
```diff
@@ -108,7 +108,7 @@ public void testIdempotence() {
     assertEquals(MaintenanceEventType.cachedValues().size(), distinctEvents.size());
 
-    // 3. Test idempotence.
+    // 3. TesConcurrencyConfigTestt idempotence.
```
`TesConcurrencyConfigTestt`?

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/MaintenanceEventTest.java:
```diff
@@ -101,7 +101,7 @@ public void testAddBrokerEvent()
 
     // Expect mocks.
-    EasyMock.expect(_mockKafkaCruiseControl.config()).andReturn(_config).times(3);
+    EasyMock.expect(_mockKafkaCruiseControl.config()).andReturn(_config).atLeastOnce();
```
(applies to similar uses in this PR on this and other tests) Can we avoid dropping `times()` in tests?

efeg commented on docs/wiki/User Guide/REST-APIs.md:
```diff
@@ -584,6 +589,7 @@ Supported parameters are:
 | disable_self_healing_for                      | list      | list of anomaly types to disable self-healing                             | N/A                   | yes       |
 | enable_self_healing_for                       | list      | list of anomaly types to enable self-healing                              | N/A                   | yes       |
 | concurrent_partition_movements_per_broker     | integer   | upper bound of ongoing replica movements into/out of a broker             | N/A                   | yes       |
+| max_partition_movements_in_cluster            | integer   | upper bound of maximum replica movements going into/out of brokers across cluster   | `default_config`  | yes |                                                            
```
(Applies to other similar updates in tables) Can we ensure that the table is formatted properly by adding spaces to ensure that `|` characters are aligned vertically?

efeg reviewed (commented):
@smtwilio Thanks for the PR!
Left some comments.

smtwilio commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java:
```diff
@@ -322,6 +326,24 @@ private void maybeAddLeaderChangeTasks(Collection<ExecutionProposal> proposals,
    */
   public List<ExecutionTask> getInterBrokerReplicaMovementTasks(Map<Integer, Integer> readyBrokers,
                                                                 Set<TopicPartition> inProgressPartitions) {
+    return getInterBrokerReplicaMovementTasks(
+        readyBrokers, inProgressPartitions, _maxNumClusterPartitionMovements);
+  }
```
I just left it overloaded in doubt. I can remove this

smtwilio commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java:
```diff
@@ -333,21 +355,39 @@ private void maybeAddLeaderChangeTasks(Collection<ExecutionProposal> proposals,
     Set<TopicPartition> partitionsInvolved = new HashSet<>();
-    while (newTaskAdded) {
+    
+    int newInProgressPartitions = inProgressPartitions.size();
```
Sure! Changing it to `num`. 
(We discussed this name internally as well. :) Started with `calculated`, `updated`,  `computed` and settled at `new`.  )

smtwilio commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/sync/AdminRequest.java:
```diff
@@ -141,6 +141,13 @@ protected String processChangeExecutionConcurrencyRequest() {
       sb.append(String.format("Execution progress check interval is set to %dMs%n", executionProgressCheckIntervalMs));
       LOG.info("Execution progress check interval is set to: {}Ms by user.", executionProgressCheckIntervalMs);
     }
+    // 5. Change max inter-broker partition concurrency.
```
Done!

smtwilio commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/AddedOrRemovedBrokerParameters.java:
```diff
@@ -35,6 +36,7 @@
     validParameterNames.add(KAFKA_ASSIGNER_MODE_PARAM);
     validParameterNames.add(BROKER_ID_PARAM);
     validParameterNames.add(CONCURRENT_PARTITION_MOVEMENTS_PER_BROKER_PARAM);
+    validParameterNames.add(MAX_PARTITION_MOVEMENTS_IN_CLUSTER_PARAM);
```
Done!

smtwilio commented:
Addressed all comments and ready for review. Build succeeds in my machine. I'm not sure why it's failing in CircleCI. 
I couldn't find any more info in CircleCI except 
```
> Task :cruise-control:integrationTest
Picked up _JAVA_OPTIONS: -Xms512m -Xmx1g

Too long with no output (exceeded 10m0s): context deadline exceeded
```

smtwilio commented:
> Addressed all comments and ready for review. Build succeeds in my machine. I'm not sure why it's failing in CircleCI. I couldn't find any more info in CircleCI except
> 
> ```
> > Task :cruise-control:integrationTest
> Picked up _JAVA_OPTIONS: -Xms512m -Xmx1g
> 
> Too long with no output (exceeded 10m0s): context deadline exceeded
> ```

It's successful! Good to review

efeg reviewed (approved):
@smtwilio Thanks for all the updates! LGTM.
